File: src/main/java/com/baulsupp/oksocial/services/twitter/PinAuthorizationFlow.java
package com.baulsupp.oksocial.services.twitter;

import ee.schimke.oksocial.output.OutputHandler;
import java.io.IOException;
import okhttp3.OkHttpClient;

public class PinAuthorizationFlow extends TwitterAuthFlow {
  public PinAuthorizationFlow(OkHttpClient client, OutputHandler outputHandler) {
    super(client, outputHandler);
  }

  protected String promptForPin(TwitterCredentials newCredentials)
      throws IOException {
    System.err.println(
        "Authorise by entering the PIN through a web browser");

    showUserLogin(newCredentials);

    return new String(System.console().readPassword("Enter PIN: "));
  }

  public TwitterCredentials authorise(String consumerKey, String consumerSecret)
      throws IOException {
    TwitterCredentials unauthed =
        new TwitterCredentials(null, consumerKey, consumerSecret, null, "");

    TwitterCredentials requestCredentials = generateRequestToken(unauthed, "oob");

    String pin = promptForPin(requestCredentials);

    return generateAccessToken(requestCredentials, pin);
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/twitter/PinAuthorizationFlow.kt
package com.baulsupp.oksocial.services.twitter

import com.baulsupp.oksocial.output.OutputHandler
import okhttp3.OkHttpClient
import okhttp3.Response
import java.io.IOException

class PinAuthorizationFlow(client: OkHttpClient, outputHandler: OutputHandler<Response>) : TwitterAuthFlow(client, outputHandler) {

  private suspend fun promptForPin(newCredentials: TwitterCredentials): String {
    System.err.println(
        "Authorise by entering the PIN through a web browser")

    showUserLogin(newCredentials)

    // TODO move to IO pool
    return String(System.console().readPassword("Enter PIN: "))
  }

  suspend fun authorise(consumerKey: String, consumerSecret: String): TwitterCredentials {
    val unauthed = TwitterCredentials(null, consumerKey, consumerSecret, null, "")

    val requestCredentials = generateRequestToken(unauthed, "oob")

    val pin = promptForPin(requestCredentials)

    return generateAccessToken(requestCredentials, pin)
  }
}
